src/public/heatmap.js

```diff
@@ -0,0 +1,593 @@
+        var polygonRings = [];
+        var arrayDepth = this._getArrayDepth(coordsArray);
+
+        if (arrayDepth === 2) {
+            polygonRings.push(this._loadLatLngAlts(coordsArray));
+        }
+        else if (arrayDepth === 3) {
+            coordsArray.forEach(function (holeCoords) {
+                polygonRings.push(this._loadLatLngAlts(holeCoords));
+            }, this);
+        }
+        else if (coordsArray.length === 0) {
+            polygonRings = [];
+        }
+        else {
+            throw new Error("Incorrect array depth for heatmap options.polygonPoints.");
+        }
+        return polygonRings;
+    },
+
+    _loadDensityParams: function (densityParams) {
+        var isArray = Array.isArray(densityParams);
+        var stop = 0.0;
+        var radius = 0.0;
+        var gain = 0.0;
+
+        if (isArray) {
+
+            if (densityParams.length < 2) {
+                throw new Error("Expected array [<stop>, <radius>, <(optional) gain>] when parsing options.densityStops");
+            }
+            stop = densityParams[0];
+            radius = densityParams[1];
+            gain = (densityParams.length > 2) ? densityParams[2] : 1.0;
+        }
+        else {
+            if (densityParams.stop === undefined || densityParams.radius === undefined) {
+                throw new Error("Expected object {stop:<stop>, radius:<radius>, (optional) gain:<gain>} when parsing options.densityStops");
+            }
+
+            stop = densityParams.stop;
+            radius = densityParams.radius;
+            gain = densityParams.gain || 1.0;
+        }
+
+        return {
+            stop: stop,
+            radius: radius,
+            gain: gain
+        };
+    },
+
+    _loadDensityStops: function (densityStopsArray) {
+        var densityStops = [];
+        var arrayDepth = this._getArrayDepth(densityStopsArray);
+
+        if (arrayDepth === 1 && typeof densityStopsArray[0] === "number") {
+            densityStops.push(this._loadDensityParams(densityStopsArray));
+        }
+        else if (arrayDepth <= 2) {
+            densityStopsArray.forEach(function (densityStopsSet) {
+                densityStops.push(this._loadDensityParams(densityStopsSet));
+            }, this);
+        }
+        else {
+            throw new Error("Incorrect array depth for heatmap options.densityStops.");
+        }
+        return densityStops;
+    },
+
+    _loadColorStop: function (colorStopParams) {
+        var isArray = Array.isArray(colorStopParams);
+        var stop = 0.0;
+        var color = "#ffffffff";
+
+        if (isArray) {
+            if (colorStopParams.length < 2) {
+                throw new Error("Expected array [<stop>, <color>] when parsing options.colorGradient");
+            }
+            stop = colorStopParams[0];
+            color = colorStopParams[1];
+        }
+        else {
+            if (colorStopParams.stop === undefined || colorStopParams.color === undefined) {
```

Comment:
Is null a concern here?

Reply:
this and other param validation / color parsing errors addressed with b1d4275d5184de940ee288af70299441413f53cd